### PR TITLE
Conversational widget: better handling of examples

### PR DESCRIPTION
**Pull request by SBrandeis**
# TL;DR

- Support an additional shape for examples in `ConversationalWidget` (see https://github.com/huggingface/huggingface.js/blob/7495ea4e13970448219c1f6862ddc5ddbd4c81b4/packages/widgets/src/routes/%2Bpage.svelte#L51-L74). => will require an update to metadata validation as well
- Display system prompts differently
- Move code around to make the `ConversationalWidget` more sound 

## Screen recording (demo)

https://github.com/huggingface/huggingface.js/assets/33657802/f7750804-061d-49da-b003-3f13ad5dbdfb

## Screen recording (embedded in Hub)

[TODO](https://github.com/huggingface/huggingface.js/assets/33657802/215fe47b-7178-4703-858d-29e140c35b2b)

Examples in README:
```yaml
widget:
  - messages:
      - role: system
        content: You are a pirate chatbot who always responds with Arr!
      - role: user
        content: There's a llama on my lawn, how can I get rid of him?
    output:
      text: >-
        Arr! 'Tis a puzzlin' matter, me hearty! A llama on yer lawn be a rare
        sight, but I've got a plan that might help ye get rid of 'im. Ye'll need
        to gather some carrots and hay, and then lure the llama away with the
        promise of a tasty treat. Once he's gone, ye can clean up yer lawn and
        enjoy the peace and quiet once again. But beware, me hearty, for there
        may be more llamas where that one came from! Arr!
    example_title: Pirate chatbot (with output)
  - text: Please help me with my homework
    example_title: text only
```

### CODE REVIEW

File: packages/widgets/src/lib/components/InferenceWidget/widgets/ConversationalWidget/ConversationalWidget.svelte [hunk 1]
```diff
@@ -72,30 +81,47 @@
...
-		if (!inferenceClient) {
-			error = "Inference client not ready";
-			return;
+			if (shouldUpdateUrl && !messages.length) {
+				updateUrl({ text: trimmedText });
+			}
+
+			// Add user message to chat
+			messages = [...messages, { role: "user", content: trimmedText }];
+			await tick();
```
Comment (Wauplin) [inline packages/widgets/src/lib/components/InferenceWidget/widgets/ConversationalWidget/ConversationalWidget.svelte:101]: TIL

**Review by Wauplin — APPROVED** [review]
Thanks for fixing the widget @SBrandeis! Demos look very nice! :fire:  

TBH I've reviewed the PR but I'm not familiar with most of the svelte/ts code (it was a cool read though^^).

About the widget, WDYT of renaming it `ChatWidget` to be consistent with ChatMessage and WidgetExampleChatInput ?

File: packages/widgets/src/lib/components/InferenceWidget/shared/inputValidation.ts [hunk 2]
```diff
@@ -13,65 +14,76 @@ import type {
...
-export function isTextInput<TOutput>(sample: WidgetExample<TOutput>): sample is WidgetExampleTextInput<TOutput> {
-	return "text" in sample;
+function isObject(arg: unknown): arg is Record<string, unknown> {
+	return !!arg && arg?.constructor === Object;
+}
+function isStrArray(arg: unknown): arg is string[] {
+	return Array.isArray(arg) && arg.every((v) => typeof v === "string");
+}
+
+export function isTextInput<TOutput>(sample: unknown): sample is WidgetExampleTextInput<TOutput> {
```
Comment (pcuenca) [inline packages/widgets/src/lib/components/InferenceWidget/shared/inputValidation.ts:24]: Not really sure how it works, but sad that we had to go to `unknown` from `WidgetExample<TOutput>` :)

**Review by pcuenca — COMMENTED** [review]
Same as @Wauplin, I think the examples are rich and nice, but not familiar with the typing details yet :)

File: packages/widgets/src/lib/components/InferenceWidget/shared/types.ts [hunk 3]
```diff
@@ -71,8 +71,3 @@ export interface ImageSegment {
 	imgData?: ImageData;
 	bitmap?: ImageBitmap;
 }
-
-export interface ChatMessage {
```
Comment (mishig25) [inline packages/widgets/src/lib/components/InferenceWidget/shared/types.ts:75]: testing github comment

File: packages/widgets/src/lib/components/InferenceWidget/shared/WidgetOutputConvo/WidgetOutputConvo.svelte [hunk 4]
```diff
@@ -26,8 +26,10 @@
 		{#each messages as message}
 			{#if message.role === "user"}
 				<WidgetOutputConvoBubble position="right" text={message.content} />
-			{:else}
+			{:else if message.role === "assistant"}
 				<WidgetOutputConvoBubble position="left" text={message.content} />
+			{:else}
+				<WidgetOutputConvoBubble position="center" text={message.content} />
 			{/if}
 		{/each}
```
Comment (mishig25) [inline packages/widgets/src/lib/components/InferenceWidget/shared/WidgetOutputConvo/WidgetOutputConvo.svelte:34]:
```suggestion
		{#each messages as {role, content}}
            {@const position = role === "user" ? "right" : role === "assistant" ? "left" : "center"}
            <WidgetOutputConvoBubble {position} text={content} />
		{/each}
```

I think we can use the `svelte const` to make it even shorter

Comment (SBrandeis) [reply packages/widgets/src/lib/components/InferenceWidget/shared/WidgetOutputConvo/WidgetOutputConvo.svelte:34]: nice!

Comment (mishig25) [reply packages/widgets/src/lib/components/InferenceWidget/shared/WidgetOutputConvo/WidgetOutputConvo.svelte:34]: or just `<WidgetOutputConvoBubble position={role === "user" ? "right" : role === "assistant" ? "left" : "center"} text={content} /> `

**Review by mishig25 — APPROVED** [review]
lgtm !

**Comment by SBrandeis** [comment]
> About the widget, WDYT of renaming it ChatWidget

No strong opinion here!

**Review by julien-c — APPROVED** [review]
lgtm from a quick glance!
